### DISCUSSION TITLE
Is possible to have starting offset = OFFSET_STORED behavior without Kafka consumer groups?

### DISCUSSION THREAD

**Discussion by BrunoBerisso** [bytewax/bytewax#501] [Q&A] [answered]
Hello.

I'm trying to make sense of how the starting offset works in Bytewax, I initially setup my Kafka consumer with a consumer group like this:

```python
flow = Dataflow("Test Flow")
source = kafka_op.input(
        f"read {name} from kafka",
        flow,
        brokers=KAFKA_BROKERS,
        topics=name,
        starting_offset=confluent_kafka.OFFSET_STORED,
        tail=KAFKA_STOP_CONSUMING,
        add_config={
                "group.id": KAFKA_CONSUMER_GROUP,
                "enable.auto.commit": "true",
                "enable.auto.offset.store": "true"
        }
    )
```
which works fin but prints a lot of warnings coming from librdkafka 
```
%4|1723113475.263|CONFWARN|rdkafka#producer-80| [thrd:app]: Configuration property group.id is a consumer property and will be ignored by this producer instance
%4|1723113475.263|CONFWARN|rdkafka#producer-80| [thrd:app]: Configuration property enable.auto.commit is a consumer property and will be ignored by this producer instance
%4|1723113475.263|CONFWARN|rdkafka#producer-80| [thrd:app]: Configuration property enable.auto.offset.store is a consumer property and will be ignored by this producer instance
```

So, I started digging into the docs and this repo and found the explanation of the warnings in the last paragraph of this comment https://github.com/bytewax/bytewax/issues/377#issuecomment-1900030506 But what's concerning are the lines above those that say:
> If you don't care about recovery, you can set group.id and enable auto.commit. **But that means that when your dataflow restarts, you loose all the messages that were received but not processed**, loosing the at-least-once semantics the operator normally offers.

I think that is related to the explanation in the docs here https://docs.bytewax.io/stable/guide/concepts/kafka.html#kafka-and-recovery which describes that messages being consumed between `auto.commit.interval.ms` might not be reported back to the group coordinator if the dataflow crashes so they will not be retried.

This let me with some questions:
- What would be the right way to get _"at-last-once"_ semantics combined with OFFSET_STORED behavior?
- It seems that Bytewax requires one worker per partition to keep track of the offsets (also mentioned here https://github.com/bytewax/bytewax/issues/377#issuecomment-1900030506), does this means that I need to manage the recovery folders and pass them to each worker?

I'll appreciate any pointers on this.
Thanks!

**Comment by Psykopear** [comment, marked as answer]
> What would be the right way to get "at-last-once" semantics combined with OFFSET_STORED behavior?

You get "at-least-once" semantics by enabling recovery and letting bytewax handle the offset. Using "group.id", committing the offsets back to kafka and restarting from OFFSET_STORED is the standard way of letting the kafka server handle this instead. The problem with that is that if you read a kafka message, commit the offset, and then the dataflow crashes before producing whatever you need in output, the dataflow will skip that message on restart, since the kafka server can't know the message wasn't fully processed.
So the proper way to use the connector if you can't skip messages is to enable recovery and let Bytewax handle the offset.

> It seems that Bytewax requires one worker per partition to keep track of the offsets

The maximum parallelization level of the input connector is defined by the total number of partitions, that's true. But if you have less workers than partition, some workers will simply handle more than one partition (if you have more workers, some workers won't read anything). So the total number of workers only defines how many partitions each worker will have to read from. Having one worker per partition might be best in some cases, but it's absolutely not required, nor necessarily ideal.

> does this means that I need to manage the recovery folders and pass them to each worker?

It depends on how you execute the dataflow. If all the processes have access to the same folder, you can initialize the folder only once. In general you need a folder per process, and if they can't be shared between processes (if each process runs on a separate machine for example), you'll need to initialize it in each place.

**Reply by BrunoBerisso** [reply]
Thanks for the answer.

I'm not sure If I understand why Bytewax can't rely on consumer groups to keep track of the consumer offsets, I guess you could turn `enable.auto.commit` off and commit the offsets once the messages are processed? 🤔 

I'd appreciate if you could point me to any resource that sheds some light here, if it's too much into the weeds please feel free to ignore it.

Thanks again.

**Reply by davidselassie** [reply]
> I'm not sure If I understand why Bytewax can't rely on consumer groups to keep track of the consumer offsets, I guess you could turn `enable.auto.commit` off and commit the offsets once the messages are processed? 🤔

Each Bytewax step is scheduled independently and run whenever an async queue has incoming messages; there's no way to know just based on dataflow definition order that any other step has run in the general case of a cluster with network delays. Bytewax currently does not provide hooks in the source API to allow the source to know when all of the other downstream steps have successfully completed their work. That hook is what would be needed to commit the offsets to the broker once they have been processed. Doing any committing within the current source API would be "too soon" and thus could result in the input committing offsets, but a later step crashes and the system should be resuming from the previous offsets.

Although those hooks are not currently exposed in the source API, there are internal mechanisms that the Bytewax runtime uses to achieve the same thing for state snapshots. That is why having Bytewax recovery manage the Kafka partition offsets _does_ result in rewinding the offsets used to read from on resume and you get the correct behavior.

We have plans to introduce those source API hooks eventually and allow for using the broker offsets as the location to store the resume state, but the current version does not have this. It's a relatively major undertaking, though, to support the correct behavior and support all the current features of cluster execution and rescaling.
